### ISSUE TITLE
Point-wise comparison operators are not differentiable

### ISSUE THREAD

**Issue by s1ddok** [tensorflow/swift-apis#771]
I've faced an issue during porting a very simple loss function. 

```
@differentiable(wrt: predicted)
func wingLoss(predicted: Tensorf, expected: Tensorf, w: Float = 10, eps: Float = 0.02) -> Tensorf {
    let x = abs(predicted - expected)
    let center = w * log(1 + x / eps)
    let c = w - w * log(1 + w / eps)
    let wings = x - c

    let loss = Tensorf(x .< w) * center + Tensorf(x .>= w) * wings
    
    return loss.mean()
}
```

This: `Tensorf(x .< w)` and this: ` Tensorf(x .>= w)` appeared to be non-differentiable. Can somebody hint on how to implement derivatives for comparison? From my understanding it is basically zero and in one point it is undefined. I'm not sure how to implement this in code for an operator.

**Comment by t-ae** [comment]
Comparison operators returns `Tensor<Bool>`, so you can't differentiate them. I don't know if there's plan to ignore them. 
For now you can use `withoutDerivative(at:)` to avoid differentiation.
```swift
let loss = withoutDerivative(at: Tensorf(x .< w)) * center + withoutDerivative(at: Tensorf(x .>= w)) * wings
```

**Comment by s1ddok** [comment]
@t-ae thatnks for the answer! will it still make the loss function valid?

**Comment by saeta** [comment]
I did a quick check in a smaller test case and it appears to work: https://colab.research.google.com/drive/1iGQwmZV1hvpsgR_mCQ8S3rT8th7MdeZy 

<img width="1679" alt="Screen Shot 2020-03-25 at 9 01 19 AM" src="https://user-images.githubusercontent.com/1284535/77557750-5980d380-6e77-11ea-988f-698058411402.png">

Thanks @t-ae for the great answer!

**Comment by s1ddok** [comment]
@saeta I've been using 
```swift
var loss = max(sign(w - x), 0) * center
loss = loss + max(sign(x - w), 0) * wings
```

as a workaround, but I'm glad to see this is working.

Still not sure how to get comfortable with understanding of differentiation stuff in S4TF. I thought `noDerivative` automatically makes stuff non-differentiable.
